crates/matrix-store-key/src/lib.rs

```diff
@@ -0,0 +1,510 @@
+        let mut mac_key_seed = Box::new([0u8; 32]);
+
+        let mut rng = thread_rng();
+
+        encryption_key.try_fill(&mut rng)?;
+        mac_key_seed.try_fill(&mut rng)?;
+
+        Ok(Self { encryption_key, mac_key_seed })
+    }
+
+    fn encryption_key(&self) -> &ChachaKey {
+        ChachaKey::from_slice(self.encryption_key.as_slice())
+    }
+
+    fn mac_key_seed(&self) -> &MacKeySeed {
+        &self.mac_key_seed
+    }
+
+    fn get_mac_key_for_table(&self, table_name: &str) -> MacKey {
+        let mut key = MacKey(Box::new([0u8; 32]));
+        let mut output = derive_key(table_name, self.mac_key_seed());
+
+        key.0.copy_from_slice(&output);
+
+        output.zeroize();
+
+        key
+    }
+
+    fn get_nonce() -> Result<[u8; XNONCE_SIZE], RandomError> {
+        let mut nonce = [0u8; XNONCE_SIZE];
+        let mut rng = thread_rng();
+
+        nonce.try_fill(&mut rng)?;
+
+        Ok(nonce)
+    }
+}
+
+/// Version specific info for the key derivation method that is used.
+#[derive(Debug, Serialize, Deserialize, PartialEq)]
+enum KdfInfo {
+    /// The PBKDF2 to Chacha key derivation variant.
+    Pbkdf2ToChaCha20Poly1305 {
+        /// The number of PBKDF rounds that were used when deriving the store
+        /// key.
+        rounds: u32,
+        /// The salt that was used when the passphrase was expanded into a store
+        /// key.
+        kdf_salt: [u8; KDF_SALT_SIZE],
+    },
+}
+
+/// Version specific info for encryption method that is used to encrypt our
+/// store key.
+#[derive(Debug, Serialize, Deserialize, PartialEq)]
+enum CipherTextInfo {
+    /// A store key encrypted using the ChaCha20Poly1305 AEAD.
+    ChaCha20Poly1305 {
+        /// The nonce that was used to encrypt the ciphertext.
+        nonce: [u8; XNONCE_SIZE],
+        /// The encrypted store key.
+        ciphertext: Vec<u8>,
+    },
+}
+
+/// An encrypted version of our store key, this can be safely stored in a
+/// database.
+#[derive(Debug, Serialize, Deserialize, PartialEq)]
+struct EncryptedStoreKey {
+    /// Info about the key derivation method that was used to expand the
+    /// passphrase into an encryption key.
+    pub kdf_info: KdfInfo,
+    /// The ciphertext with it's accompanying additional data that is needed to
+    /// decrypt the store key.
+    pub ciphertext_info: CipherTextInfo,
+}
+
+#[cfg(test)]
+mod test {
+    use serde_json::{json, Value};
+
+    use super::{Error, StoreKey};
+
+    #[test]
+    fn generating() {
+        StoreKey::new().unwrap();
+    }
+
+    #[test]
+    fn exporting_store_key() -> Result<(), Error> {
+        let passphrase = "it's a secret to everybody";
+        let store_key = StoreKey::new()?;
+
+        let encrypted = store_key.export(passphrase)?;
+        let decrypted = StoreKey::import(passphrase, &encrypted)?;
+
+        assert_eq!(store_key.inner.encryption_key, decrypted.inner.encryption_key);
+        assert_eq!(store_key.inner.mac_key_seed, decrypted.inner.mac_key_seed);
+
```

Comment:
I'd prefer a blackbox test here over one that knows about the internals. how about also encrypting something with the first and then decrypting in the second?